File: word-break/hyunolike.java
class Solution {
    public boolean wordBreak(String s, List<String> wordDict) {
        int n = s.length();
        boolean[] dp = new boolean[n+1];
        dp[0] = true;

        Set<String> wordSet = new HashSet<>(wordDict);

        for(int i = 1; i <= n; i++) {
            for(int j = 0; j < i; j++) {
                if(dp[j] && wordSet.contains(s.substring(j, i))) {
                    dp[i] = true;
                    break;
                }
            }
        }
        return dp[n];
    }
}

File: word-break/hyunolike.ts
function wordBreak(s: string, wordDict: string[]): boolean {
    const n = s.length;
    const dp:boolean[] = Array(n+1).fill(false);
    dp[0] = true;

    const wordSet = new Set(wordDict);

    for(let i = 1; i <= n; i++) {
        for(let j = 0; j < i; j++) {
            if(dp[j] && wordSet.has(s.substring(j, i))) {
                dp[i] = true;
                break;
            }
        }
    }
    return dp[n];
};